cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/AbstractStreamWriterSizeManager.java

```diff
@@ -0,0 +1,147 @@
+
+  protected AbstractStreamWriterSizeManager(HeartbeatPublisher heartbeatPublisher,
+                                            @Named(Constants.Stream.CONTAINER_INSTANCE_ID) int instanceId) {
+    this.heartbeatPublisher = heartbeatPublisher;
+    this.instanceId = instanceId;
+    this.absoluteSizes = Maps.newConcurrentMap();
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    heartbeatPublisher.startAndWait();
+    scheduledExecutor = MoreExecutors.listeningDecorator(
+      Executors.newScheduledThreadPool(EXECUTOR_POOL_SIZE,
+                                       Threads.createDaemonThreadFactory("stream-writer-size-manager")));
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    scheduledExecutor.shutdownNow();
+
+    // Send a last heartbeat before the Stream writer dies
+    for (Map.Entry<String, Long> entry : absoluteSizes.entrySet()) {
+      try {
+        heartbeatPublisher.sendHeartbeat(entry.getKey(),
+                                         new StreamWriterHeartbeat(System.currentTimeMillis(), entry.getValue(),
+                                                                   instanceId, StreamWriterHeartbeat.Type.REGULAR))
+          .get();
+      } catch (Throwable t) {
+        LOG.error("Could not publish final heartbeat for stream {} on writer instance {}",
+                  entry.getKey(), instanceId, t);
+      }
+    }
+
+    heartbeatPublisher.stopAndWait();
+  }
+
+  @Override
+  public void received(String streamName, long dataSize) {
+    if (dataSize <= 0) {
+      return;
+    }
+
+    boolean success;
+    do {
+      Long currentSize = absoluteSizes.get(streamName);
+      if (currentSize == null) {
+        success = absoluteSizes.putIfAbsent(streamName, dataSize) == null;
+        if (success) {
+          // This thread successfully put the name of a new Stream in the map,
+          // it can schedule the heartbeats for this stream too.
+          scheduleHeartbeats(streamName);
+        }
+      } else {
+        long newSize = currentSize + dataSize;
+        success = absoluteSizes.replace(streamName, currentSize, newSize);
+      }
+    } while (!success);
+  }
+
+  protected HeartbeatPublisher getHeartbeatPublisher() {
+    return heartbeatPublisher;
+  }
+
+  protected int getInstanceId() {
+    return instanceId;
+  }
+
+  protected ConcurrentMap<String, Long> getAbsoluteSizes() {
+    return absoluteSizes;
+  }
+
+  protected ListeningScheduledExecutorService getScheduledExecutor() {
+    return scheduledExecutor;
+  }
+
+  /**
+   * Schedule publishing heartbeats for the {@code streamName}. At fixed rate, a heartbeat will be sent
+   * containing the absolute size of the files own by this stream handler and concerning the stream {@code streamName}.
+   */
+  protected void scheduleHeartbeats(final String streamName) {
+    scheduledExecutor.scheduleAtFixedRate(new Runnable() {
+      @Override
+      public void run() {
```

Comment:
It would be a good idea to catch all exceptions here and log them, otherwise the Runnable will not be scheduled after an exception is thrown.